cmd/restic/cmd_cleanup.go

```diff
@@ -0,0 +1,372 @@
+	repo, err := OpenRepository(gopts)
+	if err != nil {
+		return err
+	}
+
+	lock, err := lockRepoExclusive(repo)
+	defer unlockRepo(lock)
+	if err != nil {
+		return err
+	}
+
+	// get snapshot list
+	Verbosef("get all snapshots\n")
+	snapshots, err := restic.LoadAllSnapshots(gopts.ctx, repo)
+	if err != nil {
+		return err
+	}
+
+	Verbosef("load indexes\n")
+	err = repo.LoadIndex(gopts.ctx)
+	if err != nil {
+		return err
+	}
+
+	usedBlobs, err := getUsedBlobs(gopts, repo, snapshots)
+	if err != nil {
+		return err
+	}
+
+	return Cleanup(opts, gopts, repo, usedBlobs)
+}
+
+func getUsedBlobs(gopts GlobalOptions, repo restic.Repository, snapshots []*restic.Snapshot) (restic.BlobSet, error) {
+	ctx := gopts.ctx
+
+	Verbosef("find data that is still in use for %d snapshots\n", len(snapshots))
+
+	// usedBlobs will be used as used and seen Blobs
+	usedBlobs := restic.NewBlobSet()
+
+	bar := newProgressMax(!gopts.Quiet, uint64(len(snapshots)), "snapshots")
+	bar.Start()
+	for _, sn := range snapshots {
+		debug.Log("process snapshot %v", sn.ID())
+
+		err := restic.FindUsedBlobs(ctx, repo, *sn.Tree, usedBlobs, usedBlobs)
+		if err != nil {
+			if repo.Backend().IsNotExist(err) {
+				return nil, errors.Fatal("unable to load a tree from the repo: " + err.Error())
+			}
+
+			return nil, err
+		}
+
+		debug.Log("processed snapshot %v", sn.ID())
+		bar.Report(restic.Stat{Blobs: 1})
+	}
+	bar.Done()
+	return usedBlobs, nil
+}
+
+type packInfo struct {
+	length      uint
+	usedBlobs   uint
+	unusedBlobs uint
+	tpe         restic.BlobType
+}
+
+const numLoadWorkers = 4
+
+// Start with 4 bytes overhead per pack (len of header)
+const headerLen = 4
+
+func Cleanup(opts CleanupOptions, gopts GlobalOptions, repo restic.Repository, usedBlobs restic.BlobSet) error {
+
+	ctx := gopts.ctx
+
+	Verbosef("find packs in index and calculate used size...\n")
+
+	indexPack := make(map[restic.ID]packInfo)
+	keepBlobs := restic.NewBlobSet()
+
+	for blob := range repo.Index().Each(ctx) {
+		ip, ok := indexPack[blob.PackID]
+		if !ok {
+			ip = packInfo{length: headerLen, tpe: blob.Type, usedBlobs: 0, unusedBlobs: 0}
+		}
+		// mark mixed packs with "Invalid blob type"
+		if ip.tpe != blob.Type {
+			ip.tpe = restic.InvalidBlob
+		}
+		bh := restic.BlobHandle{ID: blob.ID, Type: blob.Type}
+		if usedBlobs.Has(bh) {
+			// overhead per blob is the entry Size of the header
+			// + crypto overhead
+			ip.length += blob.Length + pack.EntrySize + crypto.Extension
+			ip.usedBlobs++
+
+			usedBlobs.Delete(bh)
+			keepBlobs.Insert(bh)
+		} else {
+			ip.unusedBlobs++
+		}
+		// update indexPack
+		indexPack[blob.PackID] = ip
+	}
+
+	// Check if all used blobs has been found in index
+	// TODO: maybe add flag --force ?
+	if len(usedBlobs) > 0 {
+		Warnf("The following blobs are missing in the index: %v", usedBlobs)
+		return errors.New("Error: Index is not complete!")
+	}
+
+	Verbosef("repack and collect packs for deletion\n")
```

Comment:
You could assign `nil` to `usedBlobs` to prevent accidental usage later on.

Reply:
mhh. usedBlobs is no longer used after the check. Is this nil-assignment really necessary?